mono/mini/interp/interp.c

```diff
@@ -6285,6 +6289,24 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, FrameClause
 			goto exit_frame;
 		}
 
+		MINT_IN_CASE(MINT_ENTER_GC_SAFE) {
+			MONO_REQ_GC_UNSAFE_MODE
+			sp = do_icall_wrapper (frame, NULL, MINT_ICALL_P_P, sp, (gpointer) mono_threads_enter_gc_safe_region_unbalanced, FALSE);
+			MONO_REQ_GC_SAFE_MODE
+			interp_in_gc_safe_state = TRUE;
+			ip += 1;
+			MINT_IN_BREAK;
+		}
+
+		MINT_IN_CASE(MINT_EXIT_GC_SAFE) {
+			MONO_REQ_GC_SAFE_MODE
+			sp = do_icall_wrapper (frame, NULL, MINT_ICALL_PP_V, sp, (gpointer) mono_threads_exit_gc_safe_region_unbalanced, FALSE);
+			MONO_REQ_GC_UNSAFE_MODE
+			interp_in_gc_safe_state = FALSE;
```

Comment:
How does this stuff work when an exception is thrown and these frames are completely unwound without resuming here ? (this is also available for jit) Or if the native code calls back into the interpreter, will we still be in gc safe ?

Reply:
I don't think `managed-to-native` frames are supposed to be completely unwound when the exception is thrown from native code. There is an interaction between XI runtime and mono runtime: Let's say native code (ObjC code) is throwing an exception. On a managed<>native barrier the XI runtime will wrap the ObjC exception into a managed exception, use `mono_runtime_set_pending_exception` and then we would resume execution in the `managed-to-native` wrapper. The thread interruption check in the wrapper would pick up that exception from there and continues with exception handling in Mono; at this point we have transitioned to GC Unsafe already.

Another scenario would be when we have:
```
<managed #2>
<native>
<managed #1>
```
And something throws an exception in `<managed #2>`, our EH transitions exception handling on the native<>managed barrier via the `MONO_FIRST_PASS_CALLBACK_TO_NATIVE` mechanism to the XI runtime. After that, it's the same path as above.

Is there some other scenario you have in mind?

Also see:
https://github.com/xamarin/xamarin-macios/blob/ba3acd709b04dd632d7abe191b9808ad5e9d1dd5/runtime/runtime.m#L2335-L2386

(Note: There is a missing GC transition in this function, https://github.com/xamarin/xamarin-macios/pull/7036/files#diff-7063c4ba2c90067ddc44fc8a7d1d865cR2364 )

Reply:
So it would work on ios because we have that callback, but in the general scenario it wouldn't. If we have m1->n->m2 and we throw from m2 while having a catch for that exception inside m1, then we would never have updated interp_in_gc_safe_state for exit and future exception throwing inside the method would be broken. This scenario likely can't happen since these are only emitted for the special m2n wrappers, but it doesn't feel like good design to me, having the true thread state on one hand and trying to emulate it with this per frame variable. If we are going to check whether we are in gc safe or unsafe code when doing exception checkpoint, can't we just check the real state ? (getting rid of the hackish per frame variable and avoiding any bugs)

Reply:
pff, that makes the patch so much simpler 😄 

updated the PR!